Add T5 and Flan-T5 example

jbochi opened the pull request:
This is a draft PR to add support for T5 modules.

It's loading all weights, but it's not applying the relative position embeddings and it's not decoding yet.

I'll continue working on it when I have the time.

awni commented:
Whoa awesome! Let me know if you'd like to collaborate (don't want to step on your toes). O/w I will review it when ready.

jbochi commented:
If you'd like to push some fixes, go for it, I'd love that :)

I won't work on it for the next 2-4 hours, but will continue later if there's still anything left for me, haha.

jbochi commented:
relative position bias now runs, but the output of the encoder doesn't match HF's

awni commented:
Nice! Coming along! I'll have some time to take a look at this tomorrow and help debug any remaining discrepencies.

jbochi commented:
Looks like the relative buckets are wrong.

It should be:

```
[[ 0, 17, 18, 19, 20, 21, 22, 23, 24, 24],
 [ 1,  0, 17, 18, 19, 20, 21, 22, 23, 24],
 [ 2,  1,  0, 17, 18, 19, 20, 21, 22, 23],
 [ 3,  2,  1,  0, 17, 18, 19, 20, 21, 22],
 [ 4,  3,  2,  1,  0, 17, 18, 19, 20, 21],
 [ 5,  4,  3,  2,  1,  0, 17, 18, 19, 20],
 [ 6,  5,  4,  3,  2,  1,  0, 17, 18, 19],
 [ 7,  6,  5,  4,  3,  2,  1,  0, 17, 18],
 [ 8,  7,  6,  5,  4,  3,  2,  1,  0, 17],
 [ 8,  8,  7,  6,  5,  4,  3,  2,  1,  0]]
 ```

My code:

```
 relative_position array([[0, 1, 2, ..., 7, 8, 9],
       [1, 0, 1, ..., 6, 7, 8],
       [2, 1, 0, ..., 5, 6, 7],
       ...,
       [7, 6, 5, ..., 0, 1, 2],
       [8, 7, 6, ..., 1, 0, 1],
       [9, 8, 7, ..., 2, 1, 0]], dtype=int32)
```

jbochi commented:
I made some progress, but I will have to stop for today.

Feel free to take it over @awni :)

jbochi commented:
Yay. It's working.

```
$ python t5.py
Missing weights:  []

Weights ignored:  ['decoder.embed_tokens.weight', 'encoder.embed_tokens.weight']
[INFO] Generating with T5...
Input:  translate English to German: That is good.
Das ist gut.
```

awni commented:
@jbochi I added a cache and made a few stylistic changes for consistency with other examples. Make sure you pull and reconvert the model as I changed the mapping.

I think there are still a couple things to figure out:

1. how to give an offset to the position bias class to avoid extra work
2. I was having trouble with the HF `decode` stripping spaces..so I had to make it batch decode for now. Would be nice to stream it.
3. FP16 by default for the sake of memory use with larger models.
4. A README 

I might have some more time to work on this tonight, but just wanted to send an update if you are working on it soon.

jbochi commented:
Thank you for the changes, Awni!

I can work on this tomorrow morning (EST), if there is still anything left
to do.

On Sun, Dec 17, 2023, 8:39 PM Awni Hannun ***@***.***> wrote:

> @jbochi <https://github.com/jbochi> I added a cache and made a few
> stylistic changes for consistency with other examples. Make sure you pull
> and reconvert the model as I changed the mapping.
>
> I think there are still a couple things to figure out:
>
>    1. how to give an offset to the position bias class to avoid extra work
>    2. I was having trouble with the HF decode stripping spaces..so I had
>    to make it batch decode for now. Would be nice to stream it.
>    3. FP16 by default for the sake of memory use with larger models.
>    4. A README
>
> I might have some more time to work on this tonight, but just wanted to
> send an update if you are working on it soon.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ml-explore/mlx-examples/pull/113#issuecomment-1859407331>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACHO2FCTIF244TLXXOBMJ3YJ6NERAVCNFSM6AAAAABAWWFW76VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQNJZGQYDOMZTGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

jbochi commented:
Thanks for the bug fix and the position bias offset!

I made some changes:
- added a README
- streaming output
- made it easy to load other models
- added an argument to serialize weights as float16, but the output of t5-large is garbage

awni commented on t5/t5.py:
```diff
@@ -287,6 +289,10 @@ def decode(
+        if self.tie_word_embeddings:
+            # Rescale output before projecting on vocab
+            # See https://github.com/huggingface/transformers/blob/71d47f0ad498b7649f11d3a9cca3cd3585e4341f/src/transformers/models/t5/modeling_t5.py#L1766C9-L1769C71
+            y *= self.model_dim ** -0.5
```
Good find! Does this account for the whole discrency?

jbochi replied:
It does not, but I finally managed to fix it in https://github.com/ml-explore/mlx-examples/pull/113/commits/64e53e84158e5c2e712de10566395c5b8a3edf5f

jbochi replied:
Oh, btw, we have the same bug in mlx: https://github.com/ml-explore/mlx/pull/210

jbochi commented:
@awni Output finally matches HF's! PTAL

```
$ python t5.py --model t5-large --prompt "translate English to German: As much as six inches of rain could fall in the New York City region through Monday morning, and officials warned of flooding along the coast."
...
 Bis Montagmorgen könnten in der Region New York City bis zu sechs Zoll Regen fallen, und die Behörden warnten vor Überschwemmungen entlang der Küste.
 
 $ python hf_t5.py --model t5-large
Bis Montagmorgen könnten in der Region New York City bis zu sechs Zoll Regen fallen, und die Behörden warnten vor Überschwemmungen entlang der Küste.
```

awni commented:
@jbochi I think this is almost ready to go. I just updated with a few more changes. I tested FP 16 for the large model, but I'd like to verify for the 3 and 7 B models and then I think we are good. What do you think?

jbochi commented:
Sounds great! But base and large are not working for me with fp16. I'm getting `<pad><pad><pad>...`

jbochi commented:
It seems to work until the first cross attention in the decoder:

```
decoder sa: x=array([[[0.420898, 8.90625, -14.1641, ..., 0.74707, 0.519531, 1.23633]]], dtype=float16), x.min()=array(-547.5, dtype=float16), x.max()=array(125.125, dtype=float16)
decoder ca: x=array([[[nan, nan, nan, ..., nan, nan, nan]]], dtype=float16), x.min()=array(inf, dtype=float16), x.max()=array(-inf, dtype=float16)
```


diff:

```
--- a/t5/t5.py
+++ b/t5/t5.py
@@ -129,9 +129,9 @@ class MultiHeadAttention(nn.Module):
         if mask is not None:
             scores = scores + mask.astype(scores.dtype)

-        scores = mx.softmax(scores.astype(mx.float32), axis=-1).astype(scores.dtype)
+        scores = mx.softmax(scores.astype(mx.float32), axis=-1)
         values_hat = (scores @ values).transpose(0, 2, 1, 3).reshape(B, L, -1)
-        return self.out_proj(values_hat), (keys, values)
+        return self.out_proj(values_hat.astype(mx.float32)).astype(queries.dtype), (keys, values)
```

awni commented:
Yea it looks like the values in the encoder are getting really large going to `inf`. Then that's causing NaNs to pop up during the RMSNorm.

awni commented:
I pushed a fix which seems to help but it's not enough for the longer prompt it still gives a slightly weird output. This model is not very friendly for low precision, the scale of the activations is pretty astounding..  maybe it's not worth trying to make it work in fp16. (Probably also be worth checking the HF reference code if they have a solution for this)

jbochi commented:
I'm getting an error now that `mlx.core` has no attribute `clip`. Is it experimental? I have `mlx-0.0.5`

The HF implementation clamps the values in a buch of places, for example:

https://github.com/huggingface/transformers/blob/238d2e3c44366aba9dc5c770c95475765a6725cb/src/transformers/models/t5/modeling_t5.py#L707-L713

I don't mind trying float16 as a followup.

BTW, I've implemented some changes to support flan-t5 too. Do you prefer that I push them here or in a separate PR?

The main difference is a "gated" activation:

```python


class DenseActivation(nn.Module):
    def __init__(self, config: T5Config):
        super().__init__()
        mlp_dims = config.d_ff or config.d_model * 4
        self.gated = config.feed_forward_proj.startswith("gated")
        if self.gated:
            self.wi_0 = nn.Linear(config.d_model, mlp_dims, bias=False)
            self.wi_1 = nn.Linear(config.d_model, mlp_dims, bias=False)
        else:
            self.wi = nn.Linear(config.d_model, mlp_dims, bias=False)
        self.wo = nn.Linear(mlp_dims, config.d_model, bias=False)
        activation = config.feed_forward_proj.removeprefix("gated-")
        if activation == "relu":
            self.act = nn.relu
        elif activation == "gelu":
            self.act = nn.gelu
        elif activation ==  "silu":
            self.act = nn.silu
        else:
            raise ValueError(f"Unknown activation: {activation}")

    def __call__(self, x):
        if self.gated:
            hidden_act = self.act(self.wi_0(x))
            hidden_linear = self.wi_1(x)
            x = hidden_act * hidden_linear
        else:
            x = self.act(self.wi(x))
        return self.wo(x)
```

awni commented:
Ah sorry it’s in main but not official release yet.

awni commented:
Ok let’s plan for float-6 as a follow up!

awni commented:
I am on board with adding support for Flan-t5 in this PR if you have it mostly done!

jbochi commented:
Done, please take a look! I tested it on a few different models:

```
$ python convert.py --model google/flan-t5-large && python t5.py --model google/flan-t5-large \
--prompt "Answer the following yes/no question by reasoning step-by-step. Can you write a whole Haiku in a single tweet?" --temp 0

Haiku is a Japanese poem that is around 108 characters long. A tweet is a short message sent on Twitter. Therefore, the final answer is yes.

$ python convert.py --model "Grammarly/coedit-large" && python t5.py  --model "Grammarly/coedit-large"  \
--prompt "Fix grammar: this sentence have grammar mistake"

This sentence has a grammar mistake.

$ python convert.py --model t5-small && python t5.py --model t5-small \
--prompt "translate English to German: As much as six inches of rain could fall in the New York City region through Monday morning, and officials warned of flooding along the coast."

Bis zu sechs Zoll Regen könnte in der Region New York bis Montag morgen fallen, und Beamte warnten vor Überschwemmungen entlang der Küste.
```

awni commented:
Brilliant, will try to get this merged tonight!

okpatil4u commented:
> I pushed a fix which seems to help but it's not enough for the longer prompt it still gives a slightly weird output. This model is not very friendly for low precision, the scale of the activations is pretty astounding.. maybe it's not worth trying to make it work in fp16. (Probably also be worth checking the HF reference code if they have a solution for this)

This is an interesting observation. Could you please advise if there is any workaround for this issue ?

awni reviewed (approved):
@jbochi way to power through this example and get it working! It looks really great to me, so I think we are set to merge.

As a follow up if you're interested in adding FP16 support that would be very welcome I'm sure by people that have less RAM. Let me know if you plan to work on it!

jbochi commented:
@awni, thanks for all your help and the kind words! I have some time today
and I will look into fp16.

On Mon, Dec 18, 2023, 11:25 PM Awni Hannun ***@***.***> wrote:

> ***@***.**** approved this pull request.
>
> @jbochi <https://github.com/jbochi> way to power through this example and
> get it working! It looks really great to me, so I think we are set to merge.
>
> As a follow up if you're interested in adding FP16 support that would be
> very welcome I'm sure by people that have less RAM. Let me know if you plan
> to work on it!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ml-explore/mlx-examples/pull/113#pullrequestreview-1788077947>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACHO2GHMSL3G2S23SMRSQLYKEJL7AVCNFSM6AAAAABAWWFW76VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTOOBYGA3TOOJUG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
